### ISSUE TITLE
Unable to assign enum value in session to a nullable enum.

### ISSUE THREAD

**Issue by judowalker** [MarkCBB/MongoDB-ASP.NET-Session-State-Store#4]
Thanks for your quick response and updates.

I'm having an issue where I store an enum value in session 

``` c#
Session["color"] = ConsoleColor.Black;
```

but I'm unable to assign it to a nullable enum

``` c#
var color = Session.Mongo<ConsoleColor?>("color");
```

I get an InvalidCastException on this line:

``` c#
return (T)BsonTypeMapper.MapToDotNetValue(sessionObj as BsonValue);
```

I also get the same exception when storing the enum in session as a string

``` c#
Session["color"] = ConsoleColor.Black.ToString();
```

and assigning it to an enum

``` c#
var color = Session.Mongo<ConsoleColor>("color");
```

**Comment by MarkCBB** [comment]
Indeed, I've reproduced this error getting nullable enums.
I'm trying some approaches but it doesn't seems a trivial problem.
As a fast solution for your project, you could use inside the enum type a value that represents the null.
For example use: ConsoleColor c = ConsoleColor.None;
Instead of ConsoleColor? c = null;
It's a common practice to avoid nullable types.
Anyway I'll try to find out a solution to enable the serialize and the deserialize of nullable types.
I'll keep you posted of my progress.

**Comment by MarkCBB** [comment]
I've found a solution that seems that solves this problem. I've to write some test cases before than publish a new package.
Anyway, tomorrow I'll publish the Helpers file adapted for serialization of nullable enums in a git branch to allow you testing and to continue developing.

**Comment by MarkCBB** [comment]
Here you have the provisional helper file that enables the serialization with a nullable enum type. This is not the definitive file! I've to write more test cases, pass them, and also I've to fix the nullable char serialization.

https://github.com/MarkCBB/MongoDB-ASP.NET-Session-State-Store/blob/Branch_test_enum/MongoSessionStateStore/MongoSessionHelpers.cs

The scenario is the following: the deserialization from a generic Object to a nullable enum always throws an exception and the session value that we get from the mongo DB is an Object (see code snippet at the end). Could be possible to serialize this object to an int value, that always work with nullable enums, and then parse the int to a nullable enum, but as we are working with generic types this generic cast is not allowed.

The solution used in this helper file is to parse the value to a string when the nullable enum is serialized. When the value is deserialized, this string is parsed to the same enum type but not nullable, and then is parsed from the not nullable to the nullable enum. In case of getting a null value, this null value will be returned prior than make this conversions, so a null reference exception never should happens.

I'll comment here when I have this problem completely fixed and published in NuGet.

```
int intVal = 0;
Object objVal = intVal; // b is as it comes from deserialization.
ConsoleColor des1 = (ConsoleColor)intVal; // ok
ConsoleColor des2 = (ConsoleColor)objVal; // ok
ConsoleColor? des3 = (ConsoleColor?)intVal; //ok
ConsoleColor? des4 = (ConsoleColor?)objVal; //invalid cast exception
```

**Comment by judowalker** [comment]
Great. Thanks for the quick update.

**Comment by MarkCBB** [comment]
Planned to be solved in release v4.0.0.

**Comment by MarkCBB** [comment]
Fixed in v4.0.0.
